veps/sig-compute/pcihole64-proposal.md

```diff
@@ -0,0 +1,323 @@
+</controller>
+```
+
+When the feature is enabled, `virtwrap/converter` used by `virt-launcher` should
+configure the `pcihole64` on the `pcie-root` controller of an `api.Domain`
+object. The required configuration can be added as part of the conversion in
+`converter.Convert_v1_VirtualMachineInstance_To_api_Domain`.
+
+Since `converter.Convert_v1_VirtualMachineInstance_To_api_Domain` does not yet
+create `pcie-root` controllers, it must also be extended to add a basic
+`pcie-root` controller to `domain.Spec.Devices.Controllers` before configuring
+the `pcihole64` on it. This can be done before or after adding other controllers
+such as usb or scsi.
+
+To allow configuring the `pcihole64` on a `pcie-root` controller the feature
+requires an addition to `virtwrap/api/schema.go`.
+
+```golang
+type Controller struct {
+Type      string            `xml:"type,attr"`
+Index     string            `xml:"index,attr"`
+Model     string            `xml:"model,attr,omitempty"`
+Driver    *ControllerDriver `xml:"driver,omitempty"`
+Alias     *Alias            `xml:"alias,omitempty"`
+Address   *Address          `xml:"address,omitempty"`
+PCIHole64 *PCIHole64        `xml:"pcihole64,omitempty"`
+}
+
+type PCIHole64 struct {
+Value uint   `xml:",chardata"`
+Unit  string `xml:"unit,attr,omitempty"`
+}
+```
+
+## API Examples
+
+<!--
+Tangible API examples used for discussion
+-->
+
+By setting the `kubevirt.io/legacyWindowsCompat` annotation on a
+`VirtualMachine` object to `"true"` the feature is enabled. Any other value
+than `"true"` will leave the feature disabled. By default, the feature will
+also be disabled.
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  name: my-legacy-vm
+  annotations:
+    kubevirt.io/legacyWindowsCompat: "true"
+spec:
+  runStrategy: Always
+  template:
+    spec:
+      domain:
+        devices: { }
+        memory:
+          guest: 512Mi
+      terminationGracePeriodSeconds: 180
+      volumes:
+        - containerDisk:
+            image: my.registry/images/legacy-os
+          name: legacy-os
+```
+
+## Alternatives
+
+<!--
+Outline any alternative designs that have been considered
+-->
+
+### Adding a specific field to the core API
+
+The feature could be configured by adding a specific field
+`spec.domain.devices.disablePCIHole64` of type `boolean` to the core API.
+
+Instead of enabling a broader compatibility behavior, this would express the
+configuration change to a `VirtualMachine` exactly.
+
+One advantage of this approach would be the improved discoverability of the
+feature, as it would become part of the core API and its documentation.
+
+The disadvantage of this approach, however, is that users must explicitly know
+that they need to disable the 64-bit PCI hole, and that the feature cannot be
+changed independently of the core API. Adding this field to the core API could
+also be interpreted as KubeVirt caring about OS specific configuration, which
```

Comment:
I thought that it was agreed that this isn't an OS specific configurable?

Reply:
I agree with Lee, IMHO, this feels less OS specific than the annotation approach.

Reply:
I'm somewhat summarizing @vladikr's concerns here. I agree that `spec.domain.devices.disablePCIHole64` does not look to OS specific to me.

Reply:
Yes, I agree this is not OS specific. However, disabling this is only needed for supporting Windows XP and such [only when QEMU and Seabios are recent enough] - sounds a bit windows specific, and qemu/seabios specific

There are no other examples of this functionality being needed elsewhere, but we are rushing to put this in the core API.

Libvirt had a good story for it, placing it under the root controller; however, KubeVirt delegates the controller's creation to Libvirt and doesn't model it.

I would opt for an annotation at this point. However, if you insist, `spec.domain.features.pciHole64` would be a better place for it (allowing for its disabling and potentially expanding it to allow for size setting), not devices.

Reply:
> I would opt for an annotation at this point. However, if you insist, `spec.domain.features.pciHole64` would be a better place for it (allowing for its disabling and potentially expanding it to allow for size setting), not devices.

`features` makes sense as we don't model the PCI controllers etc under `devices` at the moment. I'd prefer the field to explicitly disable PCIHole64 but anything is better than an annotation IMHO.